src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs

```diff
@@ -99,13 +102,60 @@ public ModelStateDictionary ModelState
         /// Gets or sets the <see cref="ViewContext"/>.
         /// </summary>
         [Activate]
-        public ViewContext ViewContext { get; set; }
+        public ViewContext ViewContext
+        {
+            get
+            {
+                if (_viewContext == null)
+                {
+                    // This should run only for the ViewComponent unit test scenarios
+                    _viewContext = new ViewContext();
+                    _viewContext.ViewData = new ViewDataDictionary(
+                        new EmptyModelMetadataProvider(),
+                        _viewContext.ModelState);
+                }
+
+                return _viewContext;
+            }
+
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _viewContext = value;
+            }
+        }
 
         /// <summary>
         /// Gets or sets the <see cref="ViewDataDictionary"/>.
         /// </summary>
         [Activate]
-        public ViewDataDictionary ViewData { get; set; }
+        public ViewDataDictionary ViewData
+        {
+            get
+            {
+                if (_viewData == null)
+                {
+                    // This should run only for the ViewComponent unit test scenarios
+                    _viewData = new ViewDataDictionary(ViewContext.ViewData);
```

Comment:
ah, I was slightly incorrect: after activation, `ViewData` will contain a copy-on-write version of what's in `ViewContext.ViewData`.  this is a little strange because other mini-scope scenarios (e.g. `HtmlHelper` recursion and tag helper activation) start that scope with consistent `ViewContext.ViewData` and `ViewData`.  the uniqueness in the method you referenced looks weird.